Recreate the README (#5)
BradleyChatha/jcli: README.md

    * User Input that integrates with User-Defined argument binding and validation.

    * Decent support for writing and parsing ANSI text.

    * Basic but flexible Configuration Providers, used alongside Dependency Injection.

    * An ANSI-enabled text buffer, for easier and efficient control over coloured, non-uniform text output.

    * Shell utilities such as `pushLocation` and `popLocation`, synonymous with Powershell's `Push-Location` and `Pop-Location`.

* Uncategorised:

    * All individual parts of this library are intended to be reusable. Allowing you to build your own CLI core using these already-made components, if desired.

# Quick Start

*This is a brief overview, for slightly more in-depth examples please look at the fully-documented [examples](https://github.com/BradleyChatha/jcli/tree/master/examples) folder.*

## Creating a default command

The default command is the command that is ran when you don't specify the name of a named command. e.g. `mytool 60 20 --some=args` would call the default command if it exists:

```d
// inside of app.d
module app;
import jaster.cli;

@Command(null, "The default command.")
struct DefaultCommand
{

>     int onExecute()

Author: I didn't consider about using `opCall`. I did however consider using a UDA.

I fear this is mostly a bikeshedding topic though, since it's more of a user-facing design choice rather than a technical one.

I went with `onExecute` so that every command has the same name for its executor function, which while it's technically less flexible than a UDA, I don't think there's really any value (especially at this point) in letting the user use whatever name they want.

It also ties in slightly with regards to inheritance, as seen in the "inheritance" section.

It might not be the best name either (e.g. `executeCommand`, or just `execute` could've been used), but I hope it's descriptive enough.

    {
        return 0;
    }
}
```

The `@Command` is a UDA (User Defined Attribute) where the first parameter is the command's name (`null` for the default command), and the second parameter is the command's description.

All commands must define an `onExecute` function, which either returns `void`, or an `int` that will be used as the program's exit/status code.

As a side note, an initial dub project does not include the intial `module app;` shown in the example above. I've added it as we'll need the directly reference the module in a later section.

## Positional Arguments

To start off, let's make our default command take a number as a positional arg. If this number is even then return `1`, otherwise return `0`.

> Positional arguments don't have a name, and are expected to exist in a specific position within the arguments passed to your program.

Author: My wording is definitely unclear here.

What I was trying to get at is that positional args don't need a name to function, and that they only exist for the user's benefit.

I might just cut the "don't have a name" part out though, since it's relatively unimportant (and incorrect).

For example the command `mytool 60 yoyo true` would have `60` in the 0th position, `yoyo` in the 1st position, and `true` in the 2nd position:

```d
@Command(null, "The default command.")
struct DefaultCommand
{

>     @CommandPositionalArg(0, "number", "The number to check.")

Reviewer: Is it possible to inherit index from the order of `@CommandPositionalArg` members like below?
```d
@CommandPositionalArg("number", "The number to check.")
int number;

@CommandPositionalArg("text", "The number to check.")
string text;
```
So `number` has `0` index and `text` has `1`.

Author: It's completely possible.

However, having behaviour that is defined by lexical order is something I'd like to avoid. `ArgBinder` being an exception since I can't think of a decent way of handling its logic.

While there are arguments such as "How often are users going to reorder their members inside of the code?", I think it's for the best to just completely prevent that from happening in the first place, at the cost of manually incrementing a number for every positional arg.

Author: I guess so. Question now is simply what how to decide between the two types.

Maybe just `CommandAutoPositionalArg`

The first parameter is the position this argument should be at, which we define as the 0th position.

The second parameter is an optional name we can give the parameter, which is shown in the command's help text, but serves no other function.

The last parameter is simply a description.

## Registering commands

We can almost use our new command, we just need to register it first. As usual I'll provide an explanation after the code:

```d
import jaster.cli;

// This is still in app.d
int main(string[] args)
{
    auto executor = new CommandLineInterface!(app);
    const statusCode = executor.parseAndExecute(args);

    UserIO.logInfof("Program exited with status code %s", statusCode);

    return statusCode;
}
```

Our main function is defined to return an `int` (status code) while also taking in any arguments passed to us via the `args` parameter.

First, we create `executor` which is a `CommandLineInterface` instance. If you are not familiar with D, the `!(app)` construct is how we pass template parameters in D, so this would be analogous to `<app>()` in C++.

> For `CommandLineInterface` to discover commands, it must know which modules (in D, this is a code file) to look in. Remember at the start I told you to write `module app;` at the start of the file? So all `!(app)` is doing is passing the module called `app` into `CommandLineInterface`, so that it can find all our commands there.

Author: My initial thinking is "no", as if I'm not mistaken `__MODULE__` is just a string of the current module's name.

After a quick test, I was right - here's part of the error: `source\app.d(6,25): instantiated from here: CommandLineInterface!"app"`

For future reference, you can pass any amount of modules into `CommandLineInterface`, not just a single one.

Second, we call `executor.parseAndExecute(args)`, which returns a status code that we store into the variable `statusCode`. This `parseAndExecute` function will parse the arguments given to it; figure out which command to call; create an instance of that command; fill out the command's argument members, and then finally call the command's `onExecute` function.

Third, we simply log the status code to the console, so we can easily see what our program is doing.

Finally, we return the status code and let the OS/shell do whatever with it.

Your app.d file should look something like [this](https://pastebin.com/aeAGa0gY).

## Compiling and running our program (Also show help text)

Open a command prompt inside the root of your project's folder, and run the command `dub build`.

If everything went well then an executable file called `mytool` (or whatever you named your project) should've been created inside of your project's root.

First, let's have a look at the help text for our default command:

```bash
$> ./mytool --help

> Usage: DEFAULT {0/number}

Author: Ah yea, I did actually have that thought when I was thinking about that.

I was in too lazy a mood to figure out how I want `CommandLineInterface` to know the program's name.

Do we go by the executable's name? (and use https://dlang.org/library/std/file/this_exe_path.html)

Do we go by a name provided by the user? (and let them pass it to the ctor or something)

I'm open to either, I just don't know which would be best.

Description:
    The default command.

Positional Args:

>     0,number                     - The number to check.

Author: I don't even really know what I was thinking to be honest.

Glad to provide a surprise though.

Opened Issue #7 for that

> ```

Author: I'm don't believe so, as anyone who's using a command line tool should already expect `--help` to both work and exist.

I find it a bit redundant.

So we can see that the help text matches the structure of our `DefaultCommand` struct.

Next, let's try out our command!

```bash
# Even number
$> ./mytool 60
Program exited with status code 1

# Odd number
$> ./mytool 59
Program exited with status code 0

# No number (Error message doesn't look great yet)
$> ./mytool
ERROR: The following required positional arguments were not provided: ["[0] number"]
Program exited with status code -1

# Too many numbers
$> ./mytool 1 2
ERROR: Stray positional arg found: '2'
Program exited with status code -1
```

Excellent, we can see that with little to no work, our command performs as expected while rejecting invalid use cases.

## Named arguments

Right, so what if the user decides that they want us to return `0` instead of `1`, and `1` instead of `0`. In other words, reverse the output?

Well, let's make them pass in a named argument called `--mode`, which maps directly to an `enum` inside of our D code, to select what behavior we want. Did I forget to mention JCLI can do that?

```d
enum Mode
{
    normal,  // Even returns 1. Odd returns 0.

>     reversed // Even returns 0. Odd return 1.

Author: I like the idea, few points in #9 that should be addressed though.

}

@Command(null, "The default command.")
struct DefaultCommand
{
    @CommandPositionalArg(0, "number", "The number to check.")
    int number;

    @CommandNamedArg("mode", "Which mode to use.")
    Mode mode;

    int onExecute()
    {
        if(this.mode == Mode.normal)
            return number % 2 == 0 ? 1 : 0;
        else
            return number % 2;
    }
}
```

> To start off, we create an `enum` called `Mode`, and give the members `normal` and `reversed`. JCLI knows how to map strings into enums.
> 
> Next, inside of `DefaultCommand` we create a member field called `Mode mode;` that is decorated with the `CommandNamedArg` UDA.

Reviewer: ```suggestion
Inside `DefaultCommand` we create a member field called `mode` that is decorated with the `CommandNamedArg` UDA and has enum type. JCLI knows how to convert argument value into enum value.
```

The first parameter is the name of the argument, which is actually important this time as this determines what name the user has to use.

The second parameter is just the description.

Then inside of `onExecute` we just check what `mode` was set to and do stuff based off of its value.

> Time to test! Remember to run `dub build` first.
> 

Reviewer: ```suggestion
```

Let's have a quick look at the help text first, to see the changes being reflected:

```bash
$> ./mytool --help

> Usage: DEFAULT {0/number} [mode]

Reviewer: Usually square brackets are used for optional arguments (see any unix tool help).

Author: #11

Description:
    The default command.

Positional Args:
    0,number                     - The number to check.

Named Args:

>     --mode                       - Which mode to use.

Reviewer: Help text should list allowed values (`normal` and `reversed` in this case):
```
    --mode                       - Which mode to use.
                                   Allowed values:
                                     normal:
                                       Even returns 1. Odd returns 0.
                                     reversed:
                                       Even returns 0. Odd returns 1.

Author: Will be part of #9

```

And now let's test our functionality:

```bash
# JCLI supports most common argument styles.

# Even (Normal)
$> mytool 60 --mode normal
Program exited with status code 1

# Even (Reversed)
$> mytool 60 --mode=reversed
Program exited with status code 0

# Bad value for mode
$> mytool 60 --mode lol

> std.conv.ConvException@\src\phobos\std\conv.d(2817): Mode does not have a member named 'reverse'

Reviewer: First of all, error text mentions `reverse` but `lol` is passed.

Error text should be better:
```suggestion
Invalid value for argument '--mode': 'lol'
```

Author: That's mostly down to laziness. Opened #10 for it though.

# Can safely assume Odd behaves properly.

# Now, we haven't marked --mode as optional, so...
$> mytool 60
ERROR: The following required named arguments were not provided: ["mode"]
Program exited with status code -1
```

We can see that `--mode` is working as expected, however notice that in the last case, the user isn't allowed to leave out `--mode` since it's not marked as optional.

## Optional Arguments

> Optional arguments, as the name implies, are optional. Only Named arguments can be optional (technically, Positional arguments can be optional in certain use cases, but JCLI doesn't support that... yet).
> 
> Inside of D's standard library - Phobos - is the module [std.typecons](https://dlang.org/phobos/std_typecons.html) which contains a type called [Nullable](https://dlang.org/phobos/std_typecons.html#Nullable).
> 
> JCLI has special support for this type, as it is used to mark an argument as optional. This type is publicly imported by JCLI, so you don't have to import anything extra.
> 
> Anyway, all we want to do is make our `mode` argument a `Nullable`, so JCLI knows it's optional:

Reviewer: ```suggestion
JCLI supports optional arguments through standard [Nullable](https://dlang.org/phobos/std_typecons.html#Nullable) type. Note that only Named arguments can be optional for now (technically, Positional arguments can be optional in certain use cases, but it's not supported ... yet).

So to make our `mode` argument optional, we need to make it `Nullable`:
```

Technically "optional" means that it can be omitted in command line and so inside a program it might be either "not set" (e.g. `null`) or to be set to some default value so consider this use case:
```
    @CommandNamedArg("mode", "Which mode to use.")
    @DefaultValue("normal")
    Mode mode;
```
or:
```d
enum {
    @CommandArgValue("normal", "Even returns 1. Odd returns 0.")
    @DefaultValue
    normal_mode,
    @CommandArgValue("reversed", "Even returns 0. Odd returns 1.")
    reverse_mode,
}
```

Author: Unless I'm missing something, [Nullable](https://dlang.org/phobos/std_typecons.html#Nullable) supports this does it not?

`Nullable.isNull` can be used to check if the user passed in a value or not.

`Nullable.get` as explained can be to get a default value if the user didn't pass a value in themselves.

```d
@Command(null, "The default command.")
struct DefaultCommand
{
    @CommandPositionalArg(0, "number", "The number to check.")
    int number;

    @CommandNamedArg("mode", "Which mode to use.")
    Nullable!Mode mode;

    int onExecute()
    {
        if(this.mode.get(Mode.normal) == Mode.normal)
            return number % 2 == 0 ? 1 : 0;
        else
            return number % 2;
    }
}
```

> The first change is that `Mode mode;` has now become `Nullable!Mode mode`, to make it, quite literally, a `Mode` that is `Nullable`.

Reviewer: Suggestion (instead of few lines here) to remove some basic things:

There are two changes:
- Type of `mode` member becomes `Nullable!Mode`.
- `onExecute` now uses `mode.get(Mode.normal)` which returns `Mode.normal` is `--mode` option is not provided.

Author: It was mostly to point out something that might be easy to glance over. I've removed the first line as its redundant with your above suggestion, but I've kept the second line.

The other change we made is that inside of `onExecute` we now use `this.mode.get(Mode.normal)`.

The `Nullable.get` function will either return us the value stored in the `Nullable`, or if the `Nullable` is null it will return to us the value we pass to it.

So by doing `get(Mode.normal)` we're saying "Give us the value the user passed in. Or, if the user didn't pass in a value, default to `Mode.normal`".

First, let's look at the help text, as it very slightly changes for nullable arguments:

```bash
$> ./mytool --help
Usage: DEFAULT {0/number} <[mode]>

Description:
    The default command.

Positional Args:
    0,number                     - The number to check.

Named Args:
    --mode                       - Which mode to use.
```

> Notice the "Usage:" line. The `[mode]` has now become `<[mode]>` to indicate it is optional.

Reviewer: As I said earlier, `[mode]` (i.e. square brackets) means it's optional (see [mkdir](https://man7.org/linux/man-pages/man1/mkdir.1.html) for example).

$> ./mytool 60
Program exited with status code 1

# Even (Reversed)
$> ./mytool 60 --mode reversed
Program exited with status code 0
```

## Arguments with multiple names

While `--mode` is nice and descriptive, it'd be nice if we could also refer to it via `-m` wouldn't it?

Here is where the very simple concept of "patterns" comes into play. At the moment, and honestly for the foreseeable future, patterns are just strings with a pipe ('|') between each different value:

```d
@Command(null, "The default command.")
struct DefaultCommand
{
    @CommandNamedArg("mode|m", "Which mode to use.")
    Nullable!Mode mode;

    // omitted as it's unchanged...
}
```

All we've done is changed `@CommandNamedArg`'s name from `"mode"` to `"mode|m"`, which basically means that we can use *either* `--mode` or `-m` to set the mode.

You can have as many values within a pattern as you want. Named Arguments cannot have whitespace within their patterns though.

Let's do a quick test as usual:

```bash
$> ./mytool 60 -m normal
Program exited with status code 1

# JCLI even supports this weird syntax shorthand arguments sometimes use.
$> ./mytool 60 -mreversed
Program exited with status code 0

> ```

Reviewer: Please add an example of how this option appears in help text.

## Named Commands

Named commands are commands that... have a name. For example `git commit`; `dub build`; `dub init`, etc. are all named commands.

It's really easy to make a named command. Let's change our default command into a named command:

```d
// Renamed from DefaultCommand
@Command("assert|a|is even", "Asserts that a number is even.")
struct AssertCommand
{
    // ...
}
```

Basically, we just pass a pattern (yes, commands can have multiple names!) as the first parameter for the `@Command` UDA, instead of leaving it as null.

Command patterns can have spaces in them, to allow a multi-word, fluent interface for your tool.

As a bit of a difference, let's test the code first:

```bash
# We have to specify a name now. JCLI will offer suggestions!
$> ./mytool 60

> ERROR: Unknown command '60'.

Reviewer: Example of error in this case (command suggestion is omitted):
```
$ git foo
git: 'foo' is not a git command. See 'git --help'.
```

Did you mean:
    assert                       - Asserts that a number is even.

# Passing cases (all producing the same output)
$> ./mytool assert 60
$> ./mytool a 60
$> ./mytool is even 60
Program exited with status code 1

> ```

Reviewer: Generated help text will be useful

```d
@ArgBinderFunc
void <anyNameItDoesntMatter>(string arg, ref <YourOutputTypeHere> output);
```

The `arg` parameter is the raw string provided by the user, for whichever argument we're binding from.

The `output` parameter is passed by reference, and the arg binder is expected to set it to the final value that'll be passed into the command instance.

Finally, all our binder does is set the `output` parameter to a `File` that opens a file using the exact `arg` given to use by the user, and opens it in read-only mode ('r').

Arg binders need to be marked with the `@ArgBinderFunc` UDA so that the `CommandLineInterface` class can discover them. Talking about `CommandLineInterface`, it'll automatically discover any arg binder from the modules you tell it about, just like it does with commands.

Let's now create our new command:

```d
@Command("cat", "Displays the contents of a file.")
struct CatCommand
{
    @CommandPositionalArg(0, "filePath", "The path to the file to display.")
    File file;

    void onExecute()
    {
        import std.stdio : writeln;

        foreach(lineInFile; this.file.byLine())
            writeln(lineInFile);
    }
}
```

> The most important thing of note here is, notice how the `file` variable has the type `File`, and recall that our arg binder's `output` parameter also has the type `File`? This tells the arg binder that it has a function to convert the user's provided string into a `File` for us.

Author: This is an interesting flaw...

So for this particular case, you change the arg binder to open in "rw" mode, and add some code to make sure the file exists, but of course that's more a bandaid than anything for various reasons.

Another roundabout way is to declare two proxy types "ReadFile" and "WriteFile", then define two separate arg binders for each type.

I don't have any real solution to this though.

(P.S. I'll address your other list of points once I'm done with the comments I can directly reply to)

> The most important thing of note here is, notice how the `file` variable has the type `File`, and recall that our arg binder's `output` parameter also has the type `File`? This tells the arg binder that it has a function to convert the user's provided string into a `File` for us.

Author: Like the idea, #23

It's cool and all being able to very easily create arg binders, but sometimes commands will need validation logic involved.

For example, some commands might only want files with a `.json` extention, while others may not care about extentions. So putting this logic into the arg binder itself isn't overly wise.

Some arguments may need validation on the pre-arg-binded string, whereas others may need validation on the post-arg-binded value. Some may need both!

JCLI handles all of this via argument validators.

Let's start off with the first example, making sure the user only passes in files with a `.json` extention, and apply it to our `cat` command. Code first, explanation after:

```d
struct HasExtention
{
    string wantedExtention;

    bool onPreValidate(string arg, ref string errorIfFalse)
    {
        import std.algorithm : endsWith;

        // Errors only display if we return false.
        errorIfFalse = "Expected file to have extention of "~this.wantedExtention;

        return arg.endsWith(this.wantedExtention);
    }
}

@Command("cat", "Displays the contents of a file.")
struct CatCommand
{
    @CommandPositionalArg(0, "filePath", "The path to the file to display.")

>     @HasExtention(".json")

Author: Possibly...

It's something I'm going to need to think over, as I'm unsure about the benefits of something like that.

It's not like you can even reuse functions in the different validation stages, since pre validation and post validation target different types, and I don't see room for any other validation stages.

The only benefit I'm seeing would potentially be clarity, but even then a properly named validator in the current style may be all you need.

I'm too unsure about it to really consider changing over to that.

    File file;

    // omitted...
}
```

To start, we create a struct called `HasExtention` and we give it a field member called `string wantedExtention;`.

Before I continue, I want to explicitly state that this validator wants to perform validation on the raw string that the user provides (pre-arg-binded) and *not* on the final value (post-arg-binded). This is referred to as "Pre Validation". So on that note...

Next, and most importantly, we define a function that specifically called `onPreValidate` that follows the following convention:

```d

> bool onPreValidate(string arg, ref string error);

Author: Yes. Even as I was writing that part I was like "why doesn't it just return a result type?".

While arg binders also take and return output via their `output` parameter, I remember having some form of reasoning for it.

Whereas this was probably just because I was lazy/tired.

#12

Because D is wonderful, it will automatically generate a constructor for us where the first parameter sets the `wantedExtention` member. So `@HasExtention(".json")` will set the extention we want to `".json"`.

And that's literally all there is to it, let's test:

```bash
# Passing
$> ./mytool cat ./dub.json
[contents of dub.json since validation was a success]
Program exited with status code 0

# Failing
$> ./mytool cat ./.gitignore
ERROR: For positional arg 0(filePath): Expected file to have extention of .json
Program exited with status code -1
```

The other type of validation is post-arg-binded validation, which performs validation on the final value provided by an arg binder.

Let's make a validator that ensures that the file is under a certain size:

```d
struct MaxSize
{
    ulong maxSize;

    bool onValidate(File file, ref string error)
    {
        error = "File is too large.";
        return file.size() <= this.maxSize;
    }
}

@Command("cat", "Displays the contents of a file.")
struct CatCommand
{
    @CommandPositionalArg(0, "filePath", "The path to the file to display.")
    @HasExtention(".json")

>     @MaxSize(2)

Author: While it does seem to be a benefit for the `@PreValidate` and `@PostValidate` style, I'm pretty certain this can be replicated with the current system already, without losing too much useability.

```d
// Untested, but should give you the idea.
//
// Writing in a style that assumes something like issue #12 is already implemented.
struct PostValidate(Type, alias Func)
{
    static assert(is(Func == ValidationResult delegate(Type value)), "Some user friendly message here.");

    ValidationResult onValidate(Type value)
    {
        return Func(value);
    }
}
```

>     @MaxSize(2)

Author: Used as:

```d
@PostValidate!File(file => file.size() <= 2)
```

>     @MaxSize(2)

Author: Or maybe even:

```d
@PostValidate!((File file) => file.size() <= 2)
```

Since you'd be able to introspect the parameter's type

>     @MaxSize(2)

Author: May as well. #24

    File file;

    // omitted...
}
```

The convention for post-arg-binded validation is almost exactly the same as pre-arg-binded validation, it also functions in exactly the same way:

```d
bool onValidate(<TYPE_OF_VALUE_TO_VALIDATE> value, ref string error);
```

The only difference is that the first parameter isn't a `string`, but instead the type of value that this validator will work with.

Validators can have different overloads of this function if required. You can even make it a template. JCLI is fine with any of that.

We've set the max size to something really small, so we can easily test that it works:

```bash
$> ./mytool cat ./dub.json
ERROR: For positional arg 0(filePath): File is too large.
Program exited with status code -1
```

Excellent.

> **Word of warning**: Due to various reasons I won't get into, JCLI will silently skip over validators that have incorrect interfaces, so if something isn't working it's likely because JCLI has found an issue with it. Specify the version `JCLI_BinderCompilerErrors` inside of your dub.json/dub.sdl in order to try and attempt to debug why.

Author: While that is definitely a benefit (and quite a large one), I wonder if it's worth it on its own to change over to it, especially since the error checking for `ArgBinder` could be improved.

The main reason ArgBinder will silently ignore invalid validators is because [of a design choice](https://github.com/BradleyChatha/jcli/blob/master/source/jaster/cli/binder.d#L106) where `ArgBinder.bind` will filter out any UDAs it doesn't understand, so the calling code doesn't need to perform filtering.

These are the two checks that `ArgBinder.bind` uses:

```d
 enum canPreValidate(alias V) = isStruct!V && __traits(compiles, { string s; bool a = typeof(V).init.onPreValidate("", s); });
 enum canValidate(alias V)    = isStruct!V && __traits(compiles, { string s; bool a = typeof(V).init.onValidate(T.init, s); });
```

Which hides/mask certain issues, so things like [version(JCLI_BinderCompilerErrors)](https://github.com/BradleyChatha/jcli/blob/master/source/jaster/cli/binder.d#L150) exist.

But looking at it, I think the better behaviour would be to just see if the UDA has a member named `onValidate` and `onPreValidate`, and just try to call it anyway, and let the compiler dish out an error message for us.

So in other words, if the UDA has `onValidate` and/or `onPreValidate`, then recognise it as a validator regardless and just let the compiler generate an error, instead of skipping over it silently.

> **Word of warning**: Due to various reasons I won't get into, JCLI will silently skip over validators that have incorrect interfaces, so if something isn't working it's likely because JCLI has found an issue with it. Specify the version `JCLI_BinderCompilerErrors` inside of your dub.json/dub.sdl in order to try and attempt to debug why.

Author: #25

The third and final parameter is a flag called `SaveOnSuccess`, which literally does at it says on the tin. If the delegate was successful, then call
`IConfig.save` to save any changes.

Finally with this `onExecute`, we just make sure that the return value was `WasExceptionThrown.no`, which explains itself.

Alternatively you can just set `IConfig.value` to something, and then call `IConfig.save`. The `IConfig.edit` function was just supposed to be a helper
around things.

We're now onto the final part of this code which is the `PrintConfigCommand`.

Literally all it does it ask for the config to be injected, retrieves its value via `IConfig.value`, and then prints it to the screen
(D's `writeln` automatically pretty prints structs).

As I said, JCLI's built-in configuration isn't terribly fancy, and offloads the majority of the work onto third party code. But it gets the job
done when you just want something up quick and easy.

## Inheritance

JCLI supports command inheritance.

The only rules with inheritance are:

* Only concrete classes can be marked with `@Command`.

> * Concrete classes must have `onExecute` defined, either by a base class or directly.

Author: This is mostly just an issue with terminology, and my apparent lack of it.

You could have `abstract class CommandBase` which is a non-concrete class (since its abstract).

Then `class CommandA : CommandBase` which is a concrete class (but isn't final).

Then `final class CommandB : CommandA` which is also a concrete class (but is final).

I think I'm just missing the proper word for "concrete".

> * Concrete classes must have `onExecute` defined, either by a base class or directly.

Author: I don't really know *why* you'd make a fully-fledged command that inherits another fully-fledged, non-abstract command, but it's possible.

> * Concrete classes must have `onExecute` defined, either by a base class or directly.

Reviewer: I couldn't think of any inheritance usage at all but it's up to you.

> * Concrete classes must have `onExecute` defined, either by a base class or directly.

Author: I remember when I was checking out dub's code a while ago (2 years I think) that the commands in dub used inheritance.

Personally, I wouldn't really use it myself, but it seems there's a precedent for it, which is why it's supported.

1. [Overview](#overview)
2. [Features](#features)
3. ["Quick" Start/HOWTO](#quick-start)
    1. [Creating your project](#creating-your-project)
    2. [Creating a default command](#creating-a-default-command)
    3. [Positional arguments](#positional-arguments)
    4. [Registering commands](#registering-commands)
    5. [Compiling and running your program (and help text)](#compiling-and-running-our-program-also-show-help-text)
    6. [Named arguments](#named-arguments)
    7. [Optional arguments](#optional-arguments)
    8. [Arguments with multiple names](#arguments-with-multiple-names)
    9. [Named commands/subcommands](#named-commands)
    10. [User Defined argument binding](#user-defined-argument-binding)
    11. [User Defined argument validation](#user-defined-argument-validation)
    12. [Unparsed Raw Arg List](#unparsed-raw-arg-list)
    13. [Dependency Injection](#dependency-injection)
    14. [Calling a command from another command](#calling-a-command-from-another-command)
    15. [Configuration](#configuration)
4. [Versions](#versions)
5. [Contributing](#contributing)

> # Features

Author: Good idea.

I'm not too sure what you meant by adding something from cli-d.

From what I can see, the only feature JCLI doesn't support is "Combining multiple short flags". When I was making the ArgPullParser, I had to choose between combining flags and allowing spaceless short-hand values (`-mbuild` is the same as `-m build`), as I believe its ambiguous to support both? Either way, for whatever reasons I may have had, I decided to support the latter syntax instead.

Looking at it though, combining flags would be probably be more useful than the (in my opinion) ugly spaceless values. So I may or may not change that in the future.

* Bash Completion support.

* Coloured, configurable logging.

* Utilities for getting input from the user. Integrates with User-Defined argument binding + validation.

* Fluently built ANSI (stylised) text, with Windows' ANSI support automatically turned on.

* Good support for parsing ANSI-encoding text, so not only can you create ANSI text, you can also read it properly.

* Basic but flexible Configuration Providers, used alongside Dependency Injection.

* An ANSI-enabled text buffer, for easier and efficient control over coloured, non-uniform text output.

* Shell utilities such as `pushLocation` and `popLocation`, synonymous with Powershell's `Push-Location` and `Pop-Location`.

* Tested on Windows, and Ubuntu 18.04.

* All individual parts of this library are intended to be reusable. Allowing you to build your own CLI core using these already-made components, if desired.

# Quick Start

*This is a brief overview, for slightly more in-depth examples please look at the fully-documented [examples](https://github.com/BradleyChatha/jcli/tree/master/examples) folder.*

> ## Creating your project

Author: Yea, I'm way too used to talking to people who are still new to coding/really bad at following directions, so I tend to over-explain and go off topic a bit.

Install any D [compiler](https://dlang.org/download.html#dmd).

Open a command prompt and run `dub init mytool` and follow the on-screen prompts.

Run the command `dub add jcli` to add JCLI as a dependency.

Open `source/app.d`, **add the line** `module app;` at the top of the file, and we'll get started from there.

## Creating a default command

The default command is the command that is ran when you don't specify the name of a named command. e.g. `mytool 60 20 --some=args` would call the default command if it exists.

To start off with, we need to import jcli and create a minimal amount of code for our command to exist.

```d
import jaster.cli;

> @Command(null, "The default command.")

Author: I get what you mean, and I even had a similar thought in the [much older](https://github.com/BradleyChatha/jcli/blob/cdb01b50bf8c395eeb1259d0cf5231c02d8f614d/source/jaster/cli/udas.d) version of JCLI.

However, and I know this is subjective, I find UDA spam to be a bit hard to read so I'd like to try and keep things into a single UDA, especially when it comes to argument UDAs. I wouldn't like to have a different UDA style for commands, and another for arguments either - I'd like them to be relatively consistent with one another.

I do agree though that having to specify a `null` name is a bit ugly and obscure.

Maybe `@DefaultCommand("description")` or `@CommandDefault("to keep consistent with things like @CommandPositionalArg")`.

It wouldn't be overly annoying to change over to two UDAs either.

idk, I'll think about it a bit more.

struct DefaultCommand
{
    int onExecute()
    {
        return 0;
    }
}
```

The `@Command` is a UDA (User Defined Attribute) where the first parameter is the command's name (`null` for the default command), and the second parameter is the command's description.

All commands must define an `onExecute` function, which either returns `void`, or an `int` which will be used as the program's exit/status code.

> Commands can either be a struct or a class, but for now we'll use structs as they're simpler.

Author: The feature itself, or just that part of the documentation?

I've moved it to somewhere a bit more appropriate in my latest changes either way.
